### PR TITLE
gh-124153: Remove `_PyType_GetModuleByDef2` private function

### PR DESCRIPTION
**Pull request by neonene**
* Issue: gh-124153

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -5297,7 +5297,7 @@ get_base_by_token_recursive(PyTypeObject *type, void *token)
     Py_ssize_t n = PyTuple_GET_SIZE(bases);
     for (Py_ssize_t i = 0; i < n; i++) {
         PyTypeObject *base = _PyType_CAST(PyTuple_GET_ITEM(bases, i));
-        if (!_PyType_HasFeature(base, Py_TPFLAGS_HEAPTYPE)) {
+        if (!PyType_HasFeature(base, Py_TPFLAGS_HEAPTYPE)) {
```
Comment (vstinner) [inline Objects/typeobject.c]: Why do you make this change? The underscore flavor is inlined and should be faster.

> Ensure the flag is checked by a function, not directly. PyType_Check()

The two functions are supposed to return the same result. It's not the case?

Comment (neonene) [reply Objects/typeobject.c]: Both are static inline functions, and now they have the same implementation with the same binary image.  It is not surprising for the internal version to switch back to a macro, which is a concern to me.  If the internal version can be faster, I would like an internal version of `PyType_Check()`.

That said, I'm fine with leaving a `_PyType_HasFeature()` in a loop.  I changed this for consistency in the related functions.

**Comment by neonene** [comment]
When I tried replacing `PyType_GetModuleBydef()` with this in `_decimal`'s `convert_op()`, the `telco` benchmark got faster by 9%.

File: Objects/typeobject.c (hunk 1 above)
Comment (neonene) [reply Objects/typeobject.c]: The `telco` slowed down by 4% with `_PyType_HasFeature()`.  Maybe a noise, but I'll check later `PyType_GetModuleByDef()` with `PyType_HasFeature()`.

Comment (neonene) [reply Objects/typeobject.c]: `PyType_GetModuleByDef()` is slower by around 10% than expected.  It seems there are a few ways to recover it.

I'll make a new PR, taking back the change in `typeobject.c`.

**Comment by neonene** [comment]
Please review #124323 regarding the performance including `PyType_GetModuleByDef()`.

**Comment by encukou** [comment]
Looks like this removes the only call to `_PyType_GetModuleByDef2`; could you also remove the function?

**Comment by vstinner** [comment]
> Looks like this removes the only call to _PyType_GetModuleByDef2; could you also remove the function?

@neonene prepared https://github.com/python/cpython/pull/124262 but this follow-up PR cannot be merged before the code is updated to no longer call _PyType_GetModuleByDef2 ;-)

**Comment by neonene** [comment]
> @neonene prepared https://github.com/python/cpython/pull/124262 but this follow-up PR cannot be merged before the code is updated to no longer call _PyType_GetModuleByDef2 ;-)

I've port the header update from #124262 to here.  Is this what you mean?

**Comment by vstinner** [comment]
Why not removing _PyType_GetModuleByDef2() body as well?

**Comment by encukou** [comment]
To be honest I don't follow too closely the order in which the changes here should be made/merged.
I also have no way to benchmark with MSVC, so I trust you about the performance, and just review for functionality and readability :)

If this removes the last call to `_PyType_GetModuleByDef2`, and we don't plan to bring the function back, then it should be removed now. If it turns out to be needed again, we can find it in Git history.

**Review by vstinner — APPROVED** [review]
LGTM

**Pull request merged**